File: apps/android/app/src/main/java/com/exponential/app/domain/SlashCommands.kt
package com.exponential.app.domain

// EXP-724: the curated steer slash-command catalog, read straight off the
// generated domain contract (`packages/domain-contract/contract.json`
// `steerCommands`) — the SAME rows every viewer's `/` menu offers, so what a
// client can pick, the desktop publisher can execute.
//
// Commands ride the ORDINARY composer path: the picked text is sent as a
// normal steer message (`/compact …` + the usual `\r` frame) and the desktop
// recognises it by its first token. Nothing here touches the wire.
//
// Mirrors `steer::commands` (desktop), `lib/steer-commands.ts` (web) and
// ExpCore's `SlashCommands.swift` (iOS) — the matching rules below are the
// same on all four and are locked by SlashCommandsTest.

/** One catalog row (a zipped view over the contract's parallel arrays). */
data class SlashCommand(
    val name: String,
    val description: String,
    /** Empty = the command takes no argument. */
    val argHint: String,
    /** The `coding_agent` values this command applies to. */
    val agents: List<String>,
    /** The client confirms before sending (context is discarded). */
    val confirm: Boolean,
) {
    /** What accepting the row puts in the composer: a trailing space only when
     *  there is an argument to type. Never auto-sends. */
    val insertion: String get() = if (argHint.isEmpty()) "/$name" else "/$name "
}

object SlashCommands {

    /** Every row, agent-agnostic, in contract order. */
    val all: List<SlashCommand> = DomainContract.steerCommandNames.indices.map { i ->
        SlashCommand(
            name = DomainContract.steerCommandNames[i],
            description = DomainContract.steerCommandDescriptions[i],
            argHint = DomainContract.steerCommandArgHints[i],
            agents = DomainContract.steerCommandAgents[i].split(",").filter { it.isNotBlank() },
            confirm = DomainContract.steerCommandConfirm[i],
        )
    }

    /** A session with no recorded agent ran on the default one. */
    private val defaultAgent: String = DomainContract.codingAgentValues.first()

    /**
     * EXP-746: the catalog id of an EXTERNAL ACP agent — deliberately a value
     * contract `codingAgent` (and so `steerCommands`) cannot name, so its
     * curated catalog is EMPTY, exactly like the desktop's
     * `steer::commands::agent_id(SessionAgent::External)`.
     */
    const val EXTERNAL_AGENT: String = "external"

    /**
     * EXP-746: which catalog a session's `/` menu and command rows key off.
     *
     * A row that names no agent is normally a claude run that predates the
     * column — but an EXTERNAL agent syncs no agent EITHER, because
     * `coding_sessions.agent` takes contract values only and there is none for
     * one. [acp] tells them apart: only the ACP engine publishes a
     * `config_state`, and an ACP run for claude/codex always stamps its id.
     * Without this the phone offered `/compact` and `/clear` — confirm dialog
     * and all — for a run whose desktop-side catalog is empty, and the literal
     * text reached the agent as a prompt. Mirrored ×4 (web `steerAgentId`, iOS
     * `SlashCommands.agentId`, desktop `slash_commands::agent_of`).
     */
    fun agentId(agent: String?, acp: Boolean): String =
        agent?.takeIf { it.isNotBlank() }?.trim()
            ?: if (acp) EXTERNAL_AGENT else defaultAgent

    /** The rows applicable to [agent] (null = the default agent), in contract
     *  order. */
    fun catalogFor(agent: String?): List<SlashCommand> {
        val id = agent?.takeIf { it.isNotBlank() } ?: defaultAgent
        return all.filter { id in it.agents }
    }

    /**
     * EXP-746: the `/` menu's catalog for a live ACP run — the CONTRACT rows
     * first, then the commands the agent itself advertised in `config_state`,
     * deduped by lowercased name (a contract row always wins: the desktop can
     * execute it, and the agent's own twin would shadow that). Byte-identical
     * ordering ×4 (`mergeAgentCommands` on web, iOS/desktop `merged`).
     *
     * An agent row carries no [SlashCommand.agents] filter: it came from THIS
     * run's agent, and the list handed in is already agent-scoped — nothing
     * re-filters a merged catalog.
     */
    fun merged(contract: List<SlashCommand>, agent: List<ConfigCommand>): List<SlashCommand> {
        if (agent.isEmpty()) return contract
        val seen = contract.mapTo(mutableSetOf()) { it.name.lowercase() }
        return contract + agent.mapNotNull { command ->
            // The wire form may or may not carry the sigil; the catalog never
            // does (its `insertion` adds it).
            val name = command.name.removePrefix("/").trim()
            if (name.isEmpty() || !seen.add(name.lowercase())) {
                null
            } else {
                SlashCommand(
                    name = name,
                    description = command.description,
                    argHint = command.hint.orEmpty(),
                    agents = emptyList(),
                    // Only the contract knows which commands discard context.
                    confirm = false,
                )
            }
        }
    }

    /**
     * The rows the `/` menu should offer for the CURRENT draft, or an empty
     * list when the menu must stay shut.
     *
     * The menu opens only while the WHOLE draft is a command being typed —
     * a leading `/` at position 0 and no whitespace yet — so `/` inside prose
     * or a path never pops it. The filter is a case-insensitive name PREFIX;
     * a bare `/` offers every row for the agent.
     */
    fun matches(
        draft: String,
        agent: String?,
        /** EXP-746: the run's own advertised commands, appended after the
         *  contract rows. Empty on a PTY run, which advertises none. */
        agentCommands: List<ConfigCommand> = emptyList(),
    ): List<SlashCommand> {
        if (!MENU_DRAFT.matches(draft)) return emptyList()
        val query = draft.drop(1)
        return merged(catalogFor(agent), agentCommands)
            .filter { it.name.startsWith(query, ignoreCase = true) }
    }

    /**
     * The catalog row [text] would run, if any: its first whitespace token
     * must equal `/name` (case-insensitive) for a row applicable to [agent].
     * Prose that merely mentions a command, and paths like `/api/foo`, are not
     * commands.
     */
    fun commandFor(
        text: String,
        agent: String?,
        agentCommands: List<ConfigCommand> = emptyList(),
    ): SlashCommand? {
        val trimmed = text.trim()
        if (!trimmed.startsWith("/")) return null
        val head = trimmed.drop(1).takeWhile { !it.isWhitespace() }
        if (head.isEmpty()) return null
        return merged(catalogFor(agent), agentCommands)
            .firstOrNull { it.name.equals(head, ignoreCase = true) }
    }

    /** The confirm dialog's title — byte-identical ×4. */
    fun confirmTitle(name: String): String = "Run /$name?"

    /** The confirm dialog's body — byte-identical ×4. */
    const val CONFIRM_BODY: String =
        "The agent forgets everything in this session so far. " +
            "Files in the worktree are kept."

    /** The confirm dialog's action — byte-identical ×4. */
    fun confirmButton(name: String): String = "Run /$name"

    /** The whole draft is a command being typed: `/`, then name characters
     *  only. The first space closes the menu (the argument is being typed). */
    private val MENU_DRAFT = Regex("^/[A-Za-z0-9-]*$")
}

File: apps/ios/ExpCore/Sources/SlashCommands.swift
import Foundation

// EXP-724 — the curated steer slash-command catalog and its two matching
// rules, byte-identical on all four clients (web `lib/steer-commands.ts`,
// Android `domain/SlashCommands.kt`, desktop `ui/src/slash_commands.rs`).
// Foundation only (ExpCore rule): the composer's menu and the confirm dialog
// are pure functions of the draft plus the session's agent, so ExpCoreTests
// drives every rule without a view.
//
// Commands ride the ORDINARY input frames as text (`/name args` + `\r`) — the
// desktop recognizes a catalog command by its first token and executes it per
// agent. There is no new relay frame, and nothing here decides what a command
// DOES.

/// One catalog row, zipped out of the generated contract arrays.
public struct SlashCommand: Equatable, Sendable, Identifiable {
    /// Bare name, no leading slash (`compact`).
    public let name: String
    public let description: String
    /// What the argument means, for the menu's muted hint. Empty = the command
    /// takes none, which is also what decides whether accepting it leaves a
    /// trailing space.
    public let argHint: String
    /// The agents that can run it — `coding_sessions.agent` values.
    public let agents: [String]
    /// Destructive enough to need a confirm before it goes out (`/clear`,
    /// the conversation is discarded).
    public let confirm: Bool

    public var id: String { name }

    public init(name: String, description: String, argHint: String, agents: [String], confirm: Bool) {
        self.name = name
        self.description = description
        self.argHint = argHint
        self.agents = agents
        self.confirm = confirm
    }

    /// What accepting this row puts in the composer: a trailing space when the
    /// command takes an argument (the caret lands where the argument goes),
    /// nothing extra when it does not. Accepting NEVER sends.
    public var insertion: String {
        argHint.isEmpty ? "/\(name)" : "/\(name) "
    }

    /// The token a sent message must open with to BE this command.
    public var token: String { "/\(name)" }
}

public enum SlashCommands {
    /// The whole catalog in contract order — the order the menu lists in.
    public static let all: [SlashCommand] = {
        let names = DomainContract.steerCommandNames
        let descriptions = DomainContract.steerCommandDescriptions
        let argHints = DomainContract.steerCommandArgHints
        let agents = DomainContract.steerCommandAgents
        let confirm = DomainContract.steerCommandConfirm
        return names.indices.compactMap { i -> SlashCommand? in
            guard i < descriptions.count, i < argHints.count,
                  i < agents.count, i < confirm.count else { return nil }
            return SlashCommand(
                name: names[i],
                description: descriptions[i],
                argHint: argHints[i],
                // The generated array carries one comma-joined string per row
                // (the contract generator has no nested-array emitter).
                agents: agents[i].split(separator: ",").map(String.init),
                confirm: confirm[i]
            )
        }
    }()

    /// The agent of a session that reports none — a pre-EXP-201 row, or a
    /// desktop old enough not to stamp it.
    public static var defaultAgent: String { DomainContract.codingAgentValues[0] }

    /// EXP-746: the catalog id of an EXTERNAL ACP agent. Deliberately a value
    /// contract `codingAgent` (and so `steerCommands`) cannot name, so its
    /// curated catalog is EMPTY — the desktop's
    /// `steer::commands::agent_id(SessionAgent::External)`.
    public static let externalAgent = "external"

    /// EXP-746: which catalog a session's `/` menu and command pills key off.
    ///
    /// A row that names no agent is normally a claude run that predates the
    /// column — but an EXTERNAL agent syncs no agent EITHER, because
    /// `coding_sessions.agent` takes contract values only and there is none
    /// for one. `acp` tells them apart: only the ACP engine publishes a
    /// `config_state`, and an ACP run for a contract agent always stamps its
    /// id. Without this a phone offered `/compact` and `/clear` — confirm
    /// dialog and all — for a run whose desktop-side catalog is empty, and
    /// the literal text reached the agent as a prompt. Mirrored ×4 (web
    /// `steerAgentId`, Android `SlashCommands.agentId`, desktop
    /// `slash_commands::agent_of`).
    public static func agentId(_ agent: String?, acp: Bool) -> String {
        let trimmed = agent?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if !trimmed.isEmpty { return trimmed }
        return acp ? externalAgent : defaultAgent
    }

    /// Every command the given agent can run, in catalog order.
    public static func catalog(for agent: String?) -> [SlashCommand] {
        let resolved = resolve(agent)
        return all.filter { $0.agents.contains(resolved) }
    }

    /// The menu's rows for a draft, or empty when the menu must not open.
    ///
    /// The menu opens iff the WHOLE draft is a leading slash plus a partial
    /// name — `^/[A-Za-z0-9-]*$`. So it opens on `/`, stays open through
    /// `/co`, and closes the moment an argument starts (`/compact `) or the
    /// slash is not at position 0 (`x /c`). The filter is a case-insensitive
    /// name PREFIX; an empty query lists everything the agent can run.
    public static func matches(draft: String, agent: String?) -> [SlashCommand] {
        guard let query = partialName(in: draft) else { return [] }
        let needle = query.lowercased()
        return catalog(for: agent).filter {
            needle.isEmpty || $0.name.lowercased().hasPrefix(needle)
        }
    }

    /// EXP-746: the `/` menu's catalog for a session whose agent advertises
    /// commands of its own (`config_state.commands`, ACP
    /// `available_commands_update`): the CONTRACT rows first, then the agent's
    /// extras, deduped by lowercased name — a contract name always wins, so an
    /// agent's `/compact` can never shadow the curated row and its confirm
    /// copy. `contract` is expected to be agent-filtered already
    /// (`catalog(for:)`); the extras came off THIS session's agent, so they
    /// carry no `agents` of their own and never leak into another agent's
    /// catalog. They also carry `confirm: false`: nothing an agent advertises
    /// is destructive enough for us to promise a dialog for it.
    public static func merged(
        _ contract: [SlashCommand], agent commands: [AgentConfigCommand]
    ) -> [SlashCommand] {
        var seen = Set(contract.map { $0.name.lowercased() })
        var merged = contract
        for command in commands {
            let name = command.name.trimmingCharacters(in: .whitespaces)
                .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            guard !name.isEmpty, seen.insert(name.lowercased()).inserted else { continue }
            merged.append(SlashCommand(
                name: name,
                description: command.description,
                argHint: command.hint ?? "",
                agents: [],
                confirm: false
            ))
        }
        return merged
    }

    /// `matches(draft:agent:)` over the merged catalog (EXP-746).
    public static func matches(
        draft: String, agent: String?, extra: [AgentConfigCommand]
    ) -> [SlashCommand] {
        guard let query = partialName(in: draft) else { return [] }
        let needle = query.lowercased()
        return merged(catalog(for: agent), agent: extra).filter {
            needle.isEmpty || $0.name.lowercased().hasPrefix(needle)
        }
    }

    /// The command a message about to be sent IS, if any: its first
    /// whitespace-separated token has to equal `/name` exactly
    /// (case-insensitively) and the command has to be runnable by this agent.
    /// Anything else — `/compactify`, `hello /compact`, `//compact` — is
    /// ordinary prose and goes out untouched.
    public static func command(for text: String, agent: String?) -> SlashCommand? {
        command(for: text, agent: agent, extra: [])
    }

    /// EXP-746: `command(for:agent:)` over the MERGED catalog — a command the
    /// agent advertised in `config_state` is a command here too, so a steered
    /// `/review` comes back into the transcript as a command pill instead of a
    /// prose bubble. The menu and the lookup read one catalog, exactly as they
    /// do on web (`parseSteerCommand` over `mergeAgentCommands`) and Android
    /// (`SlashCommands.commandFor(text, agent, agentCommands)`). The matching
    /// rules are unchanged: still the first whitespace token, still exact and
    /// case-insensitive, and an empty `extra` is the contract catalog.
    public static func command(
        for text: String, agent: String?, extra: [AgentConfigCommand]
    ) -> SlashCommand? {
        guard let token = text.split(whereSeparator: { $0.isWhitespace }).first else { return nil }
        let needle = token.lowercased()
        return merged(catalog(for: agent), agent: extra).first { $0.token.lowercased() == needle }
    }

    // MARK: - Confirm dialog copy (byte-identical ×4)

    public static func confirmTitle(_ command: SlashCommand) -> String {
        "Run /\(command.name)?"
    }

    public static let confirmBody =
        "The agent forgets everything in this session so far. Files in the worktree are kept."

    public static func confirmButton(_ command: SlashCommand) -> String {
        "Run /\(command.name)"
    }

    // MARK: - Internals

    private static func resolve(_ agent: String?) -> String {
        let trimmed = agent?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return trimmed.isEmpty ? defaultAgent : trimmed
    }

    /// The partial name a draft carries, or nil when the draft is not a
    /// menu-opening draft at all. `"/"` yields `""` (list everything).
    private static func partialName(in draft: String) -> String? {
        guard draft.hasPrefix("/") else { return nil }
        let rest = draft.dropFirst()
        for character in rest {
            guard character.isASCII,
                  character.isLetter || character.isNumber || character == "-"
            else { return nil }
        }
        return String(rest)
    }
}
